torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -758,6 +771,32 @@ void Reducer::finalize_backward() {
   local_used_maps_reduced_ = false;
 }
 
+void Reducer::runGradCallbackForVariable(
+    torch::autograd::Variable& variable,
+    GradCallback&& cb) {
+  auto context_ptr = rpc_context_.context_ptr.load();
+  if (context_ptr == nullptr) {
+    cb(variable.grad());
+  } else {
+    // Under distributed autograd
+    context_ptr->runGradCallbackForVariable(variable, std::move(cb));
+  }
+}
+
+void Reducer::RpcContext::set(ContextPtr&& new_context_ptr) {
+  if (!new_context_ptr) {
+    // Not under distributed autograd
+    return;
+  }
```

Comment:
Shouldn't we get rid of this block of code? Wouldn't there be a problem if we have this sequence:

1. We run the reducer under distributed autograd. At the end of distributed autograd context_ptr is still pointing to the old context.
2. We run the same reducer under local autograd. Now since the context pointer is not null, the callback would assume we are still in distributed autograd mode?

Although if we remove the block of code above, the reducer would set context_ptr to null under local autograd appropriately.

Reply:
That's a good catch. The test doesn't catch it because it runs the local one first. I'll make it run the distributed one first.